File: iOS/ProsaryTests/UkrainianPrayerContentTests.swift
import XCTest
@testable import Prosary

@MainActor
final class UkrainianPrayerContentTests: XCTestCase {
  private var savedFallback: [String] = []

  override func setUp() {
    super.setUp()
    savedFallback = LanguageCatalog.fallbackOrder
    LanguageCatalog.setFallbackOrder(["en", "he", "la"])
  }

  override func tearDown() {
    LanguageCatalog.setFallbackOrder(savedFallback)
    super.tearDown()
  }

  private func steps(_ bundle: String, variant: String? = nil) -> [RosaryStep] {
    PrayerEngine(calendar: StubLiturgicalCalendar()).buildSteps(for:
      Prayer(kind: .custom, languageCode: "uk", customDevotionId: bundle, variantId: variant))
  }

  func testBasicPrayersUseUkrainianTitlesAndSourcedBodiesOrDeclaredFallbacks() {
    let titles = ["Знак хреста", "Отче наш", "Радуйся, Маріє", "Слава Отцю",
                  "Апостольський символ віри", "Святий Боже", "Слався, Царице",
                  "Мати Відкупителя", "Радуйся, Царице небес", "Царице Неба"]
    XCTAssertEqual(titles.count, BasicPrayerCatalog.all.count)
    for (prayer, title) in zip(BasicPrayerCatalog.all, titles) {
      let step = BasicPrayerCatalog.step(for: prayer, languageCode: "uk")
      XCTAssertEqual(step.title, title, prayer.id)
      // These two existing antiphon translations have sourced titles only. The catalog
      // must preserve the established body fallback rather than invent liturgical text.
      if ["almaRedemptorisMater", "aveReginaCaelorum"].contains(prayer.id) {
        XCTAssertEqual(step.body, BasicPrayerCatalog.step(for: prayer, languageCode: "en").body)
        continue
      }
      XCTAssertEqual(PrayerTypography.script(of: step.body), .cyrillic, prayer.id)
      XCTAssertNotEqual(step.body, BasicPrayerCatalog.step(for: prayer, languageCode: "en").body)
    }
  }

  func testUkrainianDevotionsRenderTextInsteadOfUnresolvedKeys() {
    for bundle in ["divineMercyChaplet", "angelus", "oAntiphons", "sevenSorrows"] {
      let rendered = steps(bundle)
      XCTAssertFalse(rendered.isEmpty, bundle)
      XCTAssertTrue(rendered.contains { PrayerTypography.script(of: $0.body) == .cyrillic }, bundle)
      for step in rendered {
        XCTAssertNil(step.title.range(of: "^[a-z][A-Za-z0-9]*$", options: .regularExpression), step.title)
        XCTAssertNil(step.body.range(of: "^[a-z][A-Za-z0-9]*$", options: .regularExpression), step.body)
      }
    }
  }

  func testStationsUseUkrainianScriptureAndEditorialNarratives() {
    let scriptural = steps("stationsOfTheCross", variant: "scriptural")
    XCTAssertEqual(scriptural[2].title, "Ісус у Гетсиманському саду")
    XCTAssertTrue(scriptural[2].body.hasPrefix("І приходять на врочище Гетсиман"))
    XCTAssertTrue(scriptural[2].body.contains("Марко 14:32–36"))
    XCTAssertTrue(scriptural[2].isScripture)
    let traditional = steps("stationsOfTheCross", variant: "traditional")
    XCTAssertTrue(traditional[2].body.hasPrefix("Пилат не знаходить провини в Ісусі"))
    for number in 1...14 {
      let key = String(format: "station%02dBody", number)
      let ukrainian = PrayerPackStore.resolveBodyText(bundleId: "stationsOfTheCross", languageCode: "uk", key: key)
      XCTAssertEqual(PrayerTypography.script(of: ukrainian), .cyrillic, key)
      XCTAssertNotEqual(ukrainian, PrayerPackStore.resolveBodyText(bundleId: "stationsOfTheCross", languageCode: "en", key: key))
      XCTAssertTrue(traditional[number + 1].body.contains(ukrainian), key)
      XCTAssertFalse(traditional[number + 1].isScripture, key)
    }
  }

  func testMissingStationsOpeningAndClosingFollowTheSelectedFallback() {
    let traditional = steps("stationsOfTheCross", variant: "traditional")
    for key in ["stationsOpeningPrayer", "stationsClosingPrayer"] {
      let english = PrayerPackStore.resolveBodyText(bundleId: "stationsOfTheCross", languageCode: "en", key: key)
      XCTAssertNotEqual(english, key)
      XCTAssertEqual(PrayerPackStore.resolveBodyText(bundleId: "stationsOfTheCross", languageCode: "uk", key: key), english)
      XCTAssertTrue(traditional.contains { $0.body == english }, key)
    }
  }
}

File: Windows/Prosary.Tests/UkrainianPrayerContentTests.cs
using Prosary.Localization;
using Prosary.Models;
using Prosary.Services;
using Xunit;

namespace Prosary.Tests;

public class UkrainianPrayerContentTests : IClassFixture<PrayerPackLoaderFixture>, IDisposable
{
    private readonly string[] _savedFallback;
    public UkrainianPrayerContentTests(PrayerPackLoaderFixture _)
    {
        _savedFallback = AppSettings.LanguageFallbackOrder.ToArray();
        AppSettings.SetLanguageFallbackOrder(["en", "he", "la"]);
    }
    public void Dispose() => AppSettings.SetLanguageFallbackOrder(_savedFallback);

    private static IReadOnlyList<RosaryStep> Steps(string bundle, string? variant = null) =>
        PrayerEngine.BuildCustomDevotionSteps(bundle, "uk", false, MarianAntiphonOption.SalveRegina,
            variant, null, rosaryOptions: null, todaysGroup: MysteryGroup.Joyful, dayIndex: 0, isLent: false);

    [Fact]
    public void AllBasicPrayersUseTheSourcedUkrainianTitleAndBody()
    {
        string[] titles = ["Знак хреста", "Отче наш", "Радуйся, Маріє", "Слава Отцю", "Апостольський символ віри", "Святий Боже"];
        foreach (var (prayer, title) in BasicPrayerCatalog.All.Zip(titles))
        {
            var step = BasicPrayerCatalog.Step(prayer, "uk");
            Assert.Equal(title, step.Title);
            Assert.Equal(PrayerTypography.Script.Cyrillic, PrayerTypography.ScriptOf(step.Body));
            Assert.NotEqual(BasicPrayerCatalog.Step(prayer, "en").Body, step.Body);
        }
    }

    [Fact]
    public void UkrainianDevotionsRenderTextInsteadOfUnresolvedKeys()
    {
        foreach (var bundle in new[] { "divineMercyChaplet", "angelus", "oAntiphons", "sevenSorrows" })
        {
            var rendered = Steps(bundle);
            Assert.NotEmpty(rendered);
            Assert.Contains(rendered, step => PrayerTypography.ScriptOf(step.Body) == PrayerTypography.Script.Cyrillic);
            foreach (var step in rendered)
            {
                Assert.DoesNotMatch("^[a-z][A-Za-z0-9]*$", step.Title);
                Assert.DoesNotMatch("^[a-z][A-Za-z0-9]*$", step.Body);
            }
        }
    }

    [Fact]
    public void StationsUseUkrainianScriptureAndEditorialNarratives()
    {
        var scriptural = Steps("stationsOfTheCross", "scriptural");
        Assert.Equal("Ісус у Гетсиманському саду", scriptural[2].Title);
        Assert.StartsWith("І приходять на врочище Гетсиман", scriptural[2].Body);
        Assert.Contains("Марко 14:32–36", scriptural[2].Body);
        Assert.True(scriptural[2].IsScripture);
        var traditional = Steps("stationsOfTheCross", "traditional");
        Assert.StartsWith("Пилат не знаходить провини в Ісусі", traditional[2].Body);
        for (var number = 1; number <= 14; number++)
        {
            var key = $"station{number:D2}Body";
            var ukrainian = PrayerPackStore.ResolveBodyText("stationsOfTheCross", "uk", key);
            Assert.Equal(PrayerTypography.Script.Cyrillic, PrayerTypography.ScriptOf(ukrainian));
            Assert.NotEqual(PrayerPackStore.ResolveBodyText("stationsOfTheCross", "en", key), ukrainian);
            Assert.Contains(ukrainian, traditional[number + 1].Body);
            Assert.False(traditional[number + 1].IsScripture);
        }
    }

    [Fact]
    public void MissingStationsOpeningAndClosingFollowTheSelectedFallback()
    {
        var traditional = Steps("stationsOfTheCross", "traditional");
        foreach (var key in new[] { "stationsOpeningPrayer", "stationsClosingPrayer" })
        {
            var english = PrayerPackStore.ResolveBodyText("stationsOfTheCross", "en", key);
            Assert.NotEqual(key, english);
            Assert.Equal(english, PrayerPackStore.ResolveBodyText("stationsOfTheCross", "uk", key));
            Assert.Contains(traditional, step => step.Body == english);
        }
    }
}
